### ISSUE TITLE
solc-js throws a massive error on unhandled promise rejection

### ISSUE THREAD

**Issue by smartcontracts** [ethereum/solidity#12228]
`solc-js` throws a massive error on unhandled promise rejections. Here's a simple script you can use to reproduce:

```js
const solc = require('solc')

const main = async () => {
  throw new Error('asdf')
}

main()
```

Running this script will dump entire source of `soljson.js` to your terminal. `soljson.js` throws the following `RuntimeError`:

```
RuntimeError: abort(Error: asdf). Build with -s ASSERTIONS=1 for more info.
    at process.abort (/home/k/Projects/work/random/solcjs-test/node_modules/solc/soljson.js:1:13012
```

NodeJS wants to show you the line on which the error was triggered but `soljson.js` is a minified file where the entire source is on a single line. As a result, it prints out the entire source of `soljson.js`. Upon cleaning up the file with `js-beautify` I no longer get the entire source dumped and get the following useful error:

```
k@kpc:~/Projects/work/random/solcjs-test$ node index.js 
Error: asdf
/home/k/Projects/work/random/solcjs-test/node_modules/solc/soljson.js:622
    throw e
    ^

RuntimeError: abort(Error: asdf). Build with -s ASSERTIONS=1 for more info.
    at process.abort (/home/k/Projects/work/random/solcjs-test/node_modules/solc/soljson.js:621:13)
    at process.emit (events.js:400:28)
    at processPromiseRejections (internal/process/promises.js:245:33)
    at processTicksAndRejections (internal/process/task_queues.js:96:32)
```

This error comes from the following function inside of `soljson.js`:

```js
function abort(what) {
    if (Module["onAbort"]) {
        Module["onAbort"](what)
    }
    what += "";
    err(what);
    ABORT = true;
    EXITSTATUS = 1;
    what = "abort(" + what + "). Build with -s ASSERTIONS=1 for more info.";
    var e = new WebAssembly.RuntimeError(what);
    throw e
}
```

Which is triggered because of this listener:

```js
    process["on"]("unhandledRejection", abort);
```

I've found various workarounds to this problem. One relatively easy workaround is to remove the listener:

```
const listeners = process.listeners('unhandledRejection')
process.removeListener('unhandledRejection', listeners[listeners.length - 1])
```

However, it would be great to have a canonical fix to this problem.

**Comment by axic** [comment]
Perhaps #11689 could fix it, as noted by @ekpyron.

**Comment by chriseth** [comment]
Sorry, closed this because it said "duplicate"

**Comment by chriseth** [comment]
So we had this issue before, didn't we? And we added some workarounds that look similar to the last one suggested in the description.

**Comment by cameel** [comment]
Yeah, it's technically a duplicate but this one had better description. I'd close the other one but I thought it being open in solc-js might help people who have this issue find the workaround. I'd be fine with just closing it though.

> So we had this issue before, didn't we? And we added some workarounds that look similar to the last one suggested in the description.

We did - https://github.com/ethereum/solc-js/issues/505 was reported in December 2020 - but I don't think we did anything in response to that. I can reproduce the problem on 0.8.9 which means it has no workaround.

**Comment by aniketpcsync** [comment]
Hey, you can take a look at this answer https://github.com/trufflesuite/ganache/issues/2289 # #

**Comment by cameel** [comment]
@aniketpcsync What about it? It looks like an unrelated issue with similar effects. I.e. printing the whole file as a single line in traceback.

**Comment by cameel** [comment]
Ah, wait, it was not a question. So you're suggesting to use source maps? I'm not even sure we generate them. If we do, we discard them immediately. And I also think that disabling the minification would be much better than minifying and then having to ship source maps that negate all the speed and size benefits.

**Comment by NunoFilipeSantos** [comment]
Hi @smartcontracts! 👋 We believe this issue is fixed on our newest versions, can you try to reproduce it?

**Comment by smartcontracts** [comment]
> Hi @smartcontracts! wave We believe this issue is fixed on our newest versions, can you try to reproduce it?

Nice! Can confirm this is fixed in versions >=0.8.12

**Comment by NunoFilipeSantos** [comment]
Thank you for confirming.  Cheers! 🙌
